File: leetcode/015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  var ret = [];
  nums.sort(function(a, b) { return a - b; });
  var a, b, c;
  for (var i = 0; i < nums.length - 2; ++i) {
    if (i > 0 && a == nums[i]) continue;
    a = nums[i];

    for (var j = i + 1, k = nums.length - 1; j < k; ) {
      var s = nums[i] + nums[j] + nums[k];
      if (0 == s) {
        ret.push([nums[i], nums[j], nums[k]]);
        b = nums[j];
        while (j < k && b == nums[j]) ++j;
        c = nums[k];
        while (j < k && c == nums[k]) --k;
      } else if (s > 0) {
        c = nums[k];
        while (j < k && c == nums[k]) --k;
      } else /* if (s < 0) */ {
        b = nums[j];
        while (j < k && b == nums[j]) ++j;
      }
    }
  }
  return ret;
};

File: leetcode/015-3sum.cpp
class Solution
{
 public:
  vector<vector<int>> threeSum(vector<int>& nums)
  {
    sort(nums.begin(), nums.end());

    const int n = nums.size();
    vector<vector<int>> ret;

    for (int i = 0; i < n; ++i) {
      if (i > 0 && nums[i] == nums[i - 1]) continue;

      int sum = nums[i];
      for (int j = i + 1, k = n - 1; j < k;) {
        if (j > i + 1 && nums[j] == nums[j - 1]) {
          ++j;
          continue;
        }

        if (k < n - 1 && nums[k] == nums[k + 1]) {
          --k;
          continue;
        }

        if (nums[j] + nums[k] + sum > 0)
          --k;
        else if (nums[j] + nums[k] + sum < 0)
          ++j;
        else
          ret.push_back({nums[i], nums[j++], nums[k--]});
      }
    }

    return ret;
  }
};
